### PR TITLE
#3871 read only keys

### PR DESCRIPTION
**Pull request by tucek**
## Basics

These points need to be fulfilled for every PR:

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [x] The buildservers are happy. If not, fix **in this order**:
  - [x] add a line in `doc/news/_preparation_next_release.md`
  - [x] reformat the code with `scripts/dev/reformat-all`
  - [x] make all unit tests pass
  - [x] fix all memleaks
- [x] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

**Comment by tucek** [comment]
#3871 implemented with commit https://github.com/ElektraInitiative/libelektra/pull/3991/commits/7f646456af15b3654ed496f5d9452674847bea0d

**Comment by tucek** [comment]
This PR's branch is based on PR #3889 (!)

**Comment by tucek** [comment]
> Which part of this PR is actually new? It seems there is a lot of overlap with other Java Binding PRs...

Yes, this is because this PR's branch is based on the branch for PR #3889, which is not yet merged and github comparing it to the upstream master by default. I have combined all changes into one commit to make it more clear: https://github.com/ElektraInitiative/libelektra/pull/3991/commits/c0c5dcdc86d3327ae0d321abb4ef6aacc2b3e476

File: doc/news/_preparation_next_release.md
Comment (markus2330) [inline doc/news/_preparation_next_release.md:135]:
```suggestion
- Removed `Key::getRef`
```

File: src/bindings/jna/libelektra/src/main/java/org/libelektra/ReadOnlyKey.java
Comment (markus2330) [inline src/bindings/jna/libelektra/src/main/java/org/libelektra/ReadOnlyKey.java:21]:
```suggestion
 *          only by definition and cannot contain binary data
```

File: src/bindings/jna/libelektra/src/main/java/org/libelektra/exception/KeyBinaryValueException.java
Comment (markus2330) [inline src/bindings/jna/libelektra/src/main/java/org/libelektra/exception/KeyBinaryValueException.java:6]:
```suggestion
 * Indicates a {@link Key key's} underlying native key value is not
```

File: src/bindings/jna/libelektra/src/main/java/org/libelektra/exception/KeySetReleasedException.java
Comment (markus2330) [inline src/bindings/jna/libelektra/src/main/java/org/libelektra/exception/KeySetReleasedException.java:8]: Do we really need our own exception for this? Why would an API user want to catch this?

**Review by markus2330 — COMMENTED** [review]
Thank you for the PR!

**Review by kodebach — COMMENTED** [review]
The changes look mostly fine to me, but I have a few concerns.

1. We need to decide what the "read-only" part of `ReadOnlyKey` means. Is it the equivalent of `const Key *` in the C API or is it a different notion of "read-only"?

    IMO a different notion of "read-only" would be better. For example, in C you cannot access metadata of a `const Key * k` in C, because `keyMeta (k)` may create the metadata keyset and then needs to store it in `k`.

   In Java we don't really have these concerns. Here we could use a more sensible approach and say: `ReadOnlyKey` allows _reading_ any part of the key, including name, value and metadata, but _not writing_ anything, no matter what the C API does underneath.

2. I think we should also create interfaces `ReadableKey` and `WriteableKey` and let `ReadOnlyKey` implement `ReadableKey`, while `Key` would also implement `WriteableKey`. Then the API becomes a bit clearer.
    
    Currently e.g. `public Optional<Key> remove (ReadOnlyKey key)` is a bit confusing. You may ask "Why must the key be read-only to remove it?", until you remember that `Key` extends `ReadOnlyKey` (also a bit confusing at first).
    
    With the interfaces, we would have `public Optional<Key> remove (ReadableKey key)`. It makes sense that the key must be readable to delete it, and it is nice that it doesn't have to be writable.

File: src/bindings/jna/libelektra/src/main/java/org/libelektra/exception/KeySetReleasedException.java
Comment (tucek) [reply src/bindings/jna/libelektra/src/main/java/org/libelektra/exception/KeySetReleasedException.java:8]: I think you are right, we could do without KeySetReleasedException and KeyReleasedException, throwind an IllegalStateException instead.

File: src/bindings/jna/libelektra/src/main/java/org/libelektra/ReadOnlyKey.java
Comment (tucek) [reply src/bindings/jna/libelektra/src/main/java/org/libelektra/ReadOnlyKey.java:21]: fixed

File: doc/news/_preparation_next_release.md
Comment (tucek) [reply doc/news/_preparation_next_release.md:135]: @markus2330 I do not understand your suggestions to remove mentioning `Key::incRef` and `Key::decRef`. All three functions previously were protected and `Key` was not final. Effectively making it part of the public Java binding API.

File: src/bindings/jna/libelektra/src/main/java/org/libelektra/exception/KeyBinaryValueException.java
Comment (tucek) [reply src/bindings/jna/libelektra/src/main/java/org/libelektra/exception/KeyBinaryValueException.java:6]: I would rather leave "of type binary"... as opposed to "of type string" in `KeyStringValueException`... IMHO the value of a key being binary means only being able to assume 2 distinctive states.

**Comment by tucek** [comment]
> The changes look mostly fine to me, but I have a few concerns.
> 
> 1. We need to decide what the "read-only" part of `ReadOnlyKey` means. Is it the equivalent of `const Key *` in the C API or is it a different notion of "read-only"?
>    IMO a different notion of "read-only" would be better. For example, in C you cannot access metadata of a `const Key * k` in C, because `keyMeta (k)` may create the metadata keyset and then needs to store it in `k`.
>    In Java we don't really have these concerns. Here we could use a more sensible approach and say: `ReadOnlyKey` allows _reading_ any part of the key, including name, value and metadata, but _not writing_ anything, no matter what the C API does underneath.

i agree.

> 2. I think we should also create interfaces `ReadableKey` and `WriteableKey` and let `ReadOnlyKey` implement `ReadableKey`, while `Key` would also implement `WriteableKey`. Then the API becomes a bit clearer.
>    Currently e.g. `public Optional<Key> remove (ReadOnlyKey key)` is a bit confusing. You may ask "Why must the key be read-only to remove it?", until you remember that `Key` extends `ReadOnlyKey` (also a bit confusing at first).
>    With the interfaces, we would have `public Optional<Key> remove (ReadableKey key)`. It makes sense that the key must be readable to delete it, and it is nice that it doesn't have to be writable.

Currently `ReadOnlyKey` (as stated in java doc) primarily intended to be used for meta keys. Therefore `ReadOnlyKey` does not provide access to meta data. I currently do not see any use for other keys being of a read only type. (Are there any C API function, apart from meta data, always returning a read only key? (= locked value?))

I agree, that splitting into interfaces sounds nice at first, but i don't think it is really useful: We would need at least `ReadableKey` (without functions for returning meta data), `WritableKey`, some internal interface (e.g.`NativePointerHolder`) declaring `getPointer`, define all key parameters as `<T extends NativePointerHolder & ReadableKey>` or `<T extends NativePointerHolder & ReadableKey>`. I also don't think we need to allow different implementations of the these interfaces. `ReadOnlyKey` and `Key` are just vehicles to represent the native key resources. Therefore i made `Key` final and as of Java 17, i would also make `ReadOnlyKey` sealed. (permitting `Key` of course).

IMHO, it might be more sensible to think about a better naming for `ReadOnlyKey` ;)

**Comment by tucek** [comment]
Feedback implemented in https://github.com/ElektraInitiative/libelektra/pull/3991/commits/7def2a54929881a14cbee4e270d5ca4dc5b9efb7

**Comment by kodebach** [comment]
> Are there any C API function, apart from meta data, always returning a read only key? (= locked value?)

I don't think the C API is important here. You could just use the `ReadableKey` interface as an indicator in some Java API (possibly just inside your application, not even as public API).

> We would need at least `ReadableKey` (without functions for returning meta data), `WritableKey`, some internal interface (e.g.`NativePointerHolder`) declaring `getPointer`

That is true, it's as easy as I thought...

> IMHO, it might be more sensible to think about a better naming for `ReadOnlyKey` ;)

Definitely. Maybe just calling the class `ReadableKey` is the solution. What bugs me most about `ReadOnlyKey` is that it implies that writing is explicitly forbidden, but `Key` (which is writable) is a subclass.

**Comment by tucek** [comment]
> > IMHO, it might be more sensible to think about a better naming for `ReadOnlyKey` ;)
> 
> Definitely. Maybe just calling the class `ReadableKey` is the solution. What bugs me most about `ReadOnlyKey` is that it implies that writing is explicitly forbidden, but `Key` (which is writable) is a subclass.

I like `ReadableKey` and will update the name and improve the docs.
Regarding `Key` being a descendant of `ReadableKey` (after the rename from `ReadOnlyKey`): I don't see a problem there, since all `Key` inherit all properties of `ReadableKey`, which reflects reality. The rename resolves the semantic conflict.

**Comment by tucek** [comment]
rename done in https://github.com/ElektraInitiative/libelektra/pull/3991/commits/bf79065b18c99fc64b88bdc5ee101466b80e3a37

As soon as PR #3889 has been merged, i would like to merge this PR as well, provided there are no further objections from your side. @markus2330 please add final comments on the two unresolved discussions above.

File: src/tools/kdb/CMakeLists.txt
Comment (mpranj) [reply src/tools/kdb/CMakeLists.txt:27]: Why does the kdb tool need to be linked with pthreads? I thought it would be enough to only link `jni` with pthreads.

Maybe @markus2330 knows as it was his suggestion?

Comment (tucek) [reply src/tools/kdb/CMakeLists.txt:27]: See #3884

File: src/bindings/jna/libelektra/src/main/java/org/libelektra/ReadableKey.java
Comment (mpranj) [inline src/bindings/jna/libelektra/src/main/java/org/libelektra/ReadableKey.java:389]: Typo or is the C API naming not suitable?
```suggestion
	public boolean isDirectlyBelow (ReadableKey other)
```

File: src/tools/kdb/CMakeLists.txt
Comment (mpranj) [reply src/tools/kdb/CMakeLists.txt:27]: It is not really explained there either.

**Review by mpranj — APPROVED** [review]
Thank you, great work! :rocket: 

Let's merge it and add any further changes in a new PR.

File: src/tools/kdb/CMakeLists.txt
Comment (tucek) [reply src/tools/kdb/CMakeLists.txt:27]: That's true, only mentioned it to give some reference, to where it comes from.

File: src/bindings/jna/libelektra/src/main/java/org/libelektra/ReadableKey.java
Comment (tucek) [reply src/bindings/jna/libelektra/src/main/java/org/libelektra/ReadableKey.java:389]: Thank you, i will include this change in PR #4002

**Comment by tucek** [comment]
@mpranj Thank you for merging this PR!

File: src/tools/kdb/CMakeLists.txt
Comment (markus2330) [reply src/tools/kdb/CMakeLists.txt:27]: Technically it is needed if any library requires it. As the libraries are depending on which plugins are in use we are in a gray area. I think I suggested it because of some segfault we had in JNI which looked like that the missing pthread could be the cause. Iirc it was actually not the case. But linking against it also does not harm.

Comment (tucek) [reply src/tools/kdb/CMakeLists.txt:27]: When I'm remembering correctly, I think it was part of the problem.

**Pull request merged**
